### ISSUE TITLE
Etso in phoenix : Expected Etso.Adapter to implement Ecto.Adapter.Storage

### ISSUE THREAD

**Issue by jean-smaug** [evadne/etso#5]
👋 Hello

I'm trying to use Etso inside my Phoenix app.

So basically I replaced the adapter inside the Repo

```elixir
defmodule MyApp.Repo do
  use Ecto.Repo,
    otp_app: :my_app,
    adapter: Etso.Adapter
end
```

But during compilation I got this error

```
** (Mix) Expected Etso.Adapter to implement Ecto.Adapter.Storage in order to create storage for MyApp.Repo
```

Thx

**Comment by Laymer** [comment]
Hello there 🙂 

Complete shot in the dark here, but looking at the [Northwind repo example](https://github.com/evadne/etso/blob/develop/test/support/northwind/repo.ex) I see the following : 

```
defmodule Northwind.Repo do
  @otp_app Mix.Project.config()[:app]
  use Ecto.Repo, otp_app: @otp_app, adapter: Etso.Adapter
end
```

where the 1st line is used to resolve the compilation path of the OTP app using a macro
(see [this comment](https://stackoverflow.com/questions/40704006/where-does-the-priv-directory-belong-in-mix-umbrella-projects#comment93539361_40704350) from @evadne :) )

Is that possibly preventing Mix from building your release ?

Cheers

**Comment by jean-smaug** [comment]
Yep you're right, this part is just resolving the name off the app.

```elixir
defmodule MyApp.Repo do
  @otp_app Mix.Project.config()[:app] # returns :my_app

  use Ecto.Repo,
    otp_app: @otp_app,
    adapter: Etso.Adapter
end
```

But even with this syntax I've got the same problem 😅

**Comment by halostatue** [comment]
Ecto.Adapter.Storage may be from a newer version of Ecto than Etso was built for initially.

**Comment by evadne** [comment]
Well it is not a bad idea to add that.

**Comment by evadne** [comment]
The tasks `Mix.Tasks.Ecto.Create` and `Mix.Tasks.Ecto.Drop` require implementation of the `Ecto.Adapter.Storage` behaviour, but since Etso uses ETS which is volatile, there is no need for that.

However since Etso might not be the only Ecto adapter used in an application, the correct path forward would be no-op storage_up and use storage_down to close down all table servers IMO.

**Comment by evadne** [comment]
@jean-smaug give this a try

https://github.com/evadne/etso/tree/feature/storage-api

Main issue is that table servers can’t be located from repo config alone. Need adapter meta which has cache and pid which is not passed to Storage API. So all 3 functions are no-ops.

**Comment by evadne** [comment]
I shall close this for now. Please reopen if required.

**Comment by kevinlang** [comment]
I got an issue similar to this, and also tried the above `feature/storage-api` fix, to no avail.

However, I found that the root issue for me was a misconfiguration on my part. After creating my new Etso repo, I added it to the `ecto_repos` in `config.exs`.

```
config :hanzi,
  ecto_repos: [MyApp.Repo, MyApp.MyEtsoThing]
```

This config is used to determine whether to run these migration tasks, which assume that there is `Storage` functionality, naturally! https://hexdocs.pm/ecto/Ecto.html#module-mix-tasks-and-generators

Removing that line fixed it for me. E.g.,
```
config :hanzi,
  ecto_repos: [MyApp.Repo]
```

So if you get a similar error, it may be because you added your repo to the `:ecto_repos`, and removing it may solve your problem.
